File: src/main/java/com/music/music_inventory_api/service/ArtistService.java
package com.music.music_inventory_api.service;

import com.music.music_inventory_api.dto.request.CreateArtistRequest;
import com.music.music_inventory_api.dto.request.UpdateArtistRequest;
import com.music.music_inventory_api.dto.response.AlbumResponse;
import com.music.music_inventory_api.dto.response.ArtistResponse;
import java.util.List;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service interface for Artist entity operations. Defines business logic for
 * managing artists.
 */
public interface ArtistService
{

    /**
     * Create a new artist.
     *
     * @param request
     *            the artist creation request
     * @return the created artist response
     */
    ArtistResponse createArtist(CreateArtistRequest request);

    /**
     * Get artist by ID.
     *
     * @param id
     *            the artist ID
     * @return the artist response
     * @throws com.music.music_inventory_api.exception.EntityNotFoundException
     *             if artist not found
     */
    ArtistResponse getArtistById(Long id);

    /**
     * Get all artists with pagination.
     *
     * @param pageable
     *            pagination information
     * @return page of artist responses
     */
    Page<ArtistResponse> getAllArtists(Pageable pageable);

    /**
     * Update an existing artist.
     *
     * @param id
     *            the artist ID
     * @param request
     *            the artist update request
     * @return the updated artist response
     * @throws com.music.music_inventory_api.exception.EntityNotFoundException
     *             if artist not found
     */
    ArtistResponse updateArtist(Long id, UpdateArtistRequest request);

    /**
     * Delete an artist by ID.
     *
     * @param id
     *            the artist ID
     * @throws com.music.music_inventory_api.exception.EntityNotFoundException
     *             if artist not found
     */
    void deleteArtist(Long id);

    /**
     * Get all albums by artist ID.
     *
     * @param artistId
     *            the artist ID
     * @return list of album responses
     * @throws com.music.music_inventory_api.exception.EntityNotFoundException
     *             if artist not found
     */
    List<AlbumResponse> getAlbumsByArtist(Long artistId);
}

File: frontend/src/services/artistService.ts
import api from './api';

export interface Artist {
    id: number;
    name: string;
    bio?: string;
    imageUrl?: string;
}

export interface CreateArtistRequest {
    name: string;
    bio?: string;
    imageUrl?: string;
}

export interface UpdateArtistRequest {
    name?: string;
    bio?: string;
    imageUrl?: string;
}

class ArtistService {
    private readonly endpoint = '/artists';

    async getAll(): Promise<Artist[]> {
        const response = await api.get<Artist[]>(this.endpoint);
        return response.data;
    }

    async getById(id: number): Promise<Artist> {
        const response = await api.get<Artist>(`${this.endpoint}/${id}`);
        return response.data;
    }

    async create(artistData: CreateArtistRequest): Promise<Artist> {
        const response = await api.post<Artist>(this.endpoint, artistData);
        return response.data;
    }

    async update(id: number, artistData: UpdateArtistRequest): Promise<Artist> {
        const response = await api.put<Artist>(`${this.endpoint}/${id}`, artistData);
        return response.data;
    }

    async delete(id: number): Promise<void> {
        await api.delete(`${this.endpoint}/${id}`);
    }
}

export default new ArtistService();
